### PR TITLE
Flatten repositories from JEP-229 plugins

### PR DESCRIPTION
**Pull request by Vlatombe**
This workarounds https://github.com/mojohaus/flatten-maven-plugin/issues/362

and strips `<repositories>` from deployed pom for plugins using JEP-229.

### Testing done

```[tasklist]
### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by jglick — APPROVED** [review]
Would also be great to have an IT covering the basic build cycle of a JEP-229-based plugin unlike https://github.com/jenkinsci/plugin-pom/blob/e46655d2c7bbe610edc0cc5d05eb35486ad55210/src/it/incrementals-and-plugin-bom/.mvn/maven.config#L1-L2 and the validation script could do some stuff like assert that the top-level POM elements are only those in a specified list.

File: pom.xml [hunk 1]
```diff
@@ -1338,6 +1338,9 @@
                   <flattenMode>oss</flattenMode>
                   <outputDirectory>${project.build.directory}</outputDirectory>
                   <flattenedPomFilename>${project.artifactId}-${project.version}.pom</flattenedPomFilename>
+                  <pomElements>
```
Comment (jglick) [inline pom.xml:1341]:
```suggestion
                  <!-- TODO https://github.com/mojohaus/flatten-maven-plugin/issues/362 -->
                  <pomElements>
```

**Comment by Vlatombe** [comment]
> IT covering the basic build cycle of a JEP-229-based plugin

I'll see what I can do

File: src/it/cd-plugin/src/main/java/test/SampleRootAction.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package test;
+
+import hudson.Extension;
+import hudson.model.RootAction;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+
+@Extension
+public class SampleRootAction implements RootAction {
```
Comment (jglick) [inline src/it/cd-plugin/src/main/java/test/SampleRootAction.java:9]: Could delete this file—not adding anything AFAICT.

File: src/it/cd-plugin/src/test/java/test/SampleRootActionTest.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package test;
+
+import static org.junit.Assert.*;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class SampleRootActionTest {
```
Comment (jglick) [inline src/it/cd-plugin/src/test/java/test/SampleRootActionTest.java:10]: ditto

File: src/it/cd-plugin/src/test/resources/test/expected.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+sample served
```
Comment (jglick) [inline src/it/cd-plugin/src/test/resources/test/expected.txt:1]: ditto

File: src/it/cd-plugin/postbuild.groovy [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+assert installedPom.file
+
+assert installedPom.text.contains("<name>")
+assert installedPom.text.contains("<description>")
+assert installedPom.text.contains("<dependencies>")
+assert installedPom.text.contains("<build>")
+assert installedPom.text.contains("<scm>")
+assert !installedPom.text.contains("<properties>")
+assert !installedPom.text.contains("<dependencyManagement>")
+assert !installedPom.text.contains("<repositories>")
```
Comment (jglick) [inline src/it/cd-plugin/postbuild.groovy:20]: (the point of the PR, FTR)

File: src/it/cd-plugin/invoker.properties [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# Same options as in https://github.com/jenkins-infra/jenkins-maven-cd-action/blob/master/run.sh
```
Comment (jglick) [inline src/it/cd-plugin/invoker.properties:1]:
```suggestion
# Same options as in https://github.com/jenkins-infra/jenkins-maven-cd-action/blob/master/run.sh (but hard-coding a sample changelist)
```

**Comment by basil** [comment]
Does this need to be side-ported to https://github.com/jenkinsci/pom?

**Comment by Vlatombe** [comment]
Good point, https://github.com/jenkinsci/pom/pull/464

**Pull request merged**